File: ta_lib/strategies/signal/src/zerocross/cfo.rs
use base::prelude::*;
use core::prelude::*;
use momentum::cfo;
use timeseries::prelude::*;

pub struct CfoZeroCrossSignal {
    source_type: SourceType,
    period: usize,
}

impl CfoZeroCrossSignal {
    pub fn new(source_type: SourceType, period: f32) -> Self {
        Self {
            source_type,
            period: period as usize,
        }
    }
}

impl Signal for CfoZeroCrossSignal {
    fn lookback(&self) -> usize {
        self.period
    }

    fn trigger(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let cfo = cfo(&data.source(self.source_type), self.period);

        (cfo.cross_over(&ZERO), cfo.cross_under(&ZERO))
    }
}

File: factor/signal/zerocross/cfo.py
from dataclasses import dataclass

from core.models.parameter import (
    Parameter,
    StaticParameter,
)
from core.models.source import SourceType
from factor.signal.base import Signal, SignalType


@dataclass(frozen=True)
class CfoZeroCrossSignal(Signal):
    type: SignalType = SignalType.CfoZeroCross
    source_type: Parameter = StaticParameter(SourceType.CLOSE)
    period: Parameter = StaticParameter(14.0)
